doc/bsp/raspberrypi.md

```diff
@@ -0,0 +1,330 @@
+This document describe how to setup EMLinux for Raspberry Pi 3 model B+.
+
+## Directory structures
+
+In this document, following directory structure is used.
+
+```
+build/
+downloads/
+emlinux/repos/meta-debian 
+emlinux/repos/meta-debian-extended
+emlinux/repos/meta-emlinux
+emlinux/repos/poky
+firmware
+linux-firmware
+u-boot
+```
+
+## Build EMLinux
+
+1. Basic setup
+
+Run following command.
+
+```
+$ source repos/meta-emlinux/scripts/setup-emlinux build
+```
+
+2. Add layer
+
+EMLinux for Raspberry Pi uses additional layer, so add it.
+
+```
+$ bitbake-layers add-layer /<path to emlinux directory>/repos/meta-emlinux/bsp/meta-raspberrypi_emlinux
+```
+
+3. Build image
+
+In the build directroy, you can build image. You needs to set raspberrypi3-64 to MACHINE variable. 
+
+```
+$ MACHINE=raspberrypi3-64 bitbake core-image-minimal
+```
+
+## Build u-boot
+
+Building u-boot, you must have cross compiler. 
+
+1. Download source code
+
+```
+$ git clone https://github.com/u-boot/u-boot
+```
+
+2. Checkout release version
+
+```
+$ cd u-boot
+$ git checkout -b v201907 v2019.07
+```
+
+3. Build u-boot
+
+```
+$ make CROSS_COMPILE=aarch64-linux-gnu- rpi_3_defconfig
+$ make CROSS_COMPILE=aarch64-linux-gnu- u-boot.bin
+```
+
+## Create sdcard image
+
+This document uses parted to create partions and /dev/sde as sdcard device.
+
+1. Create fat32 partition for boot.
+
+```
+$ sudo parted /dev/sde mkpart primary fat32 1024KiB 65MiB Information: You may need to update /etc/fstab.
+```
+
+2. Create root partition
+
+```
+$ sudo parted /dev/sde mkpart primary ext4  65MiB 1GiB Information: You may need to update /etc/fstab.
+```
+
+3. Create fat32 file system on fat32 partition
+
+```
+$ sudo mkfs.vfat -n rpi -F32 /dev/sde1
+mkfs.fat 4.1 (2017-01-24)
+mkfs.fat: warning - lowercase labels might not work properly with DOS or Windows
+```
+_note: command line adds rpi label_
+
+4. Create file ext4 file system on ext4 partition
+
+```
+$ sudo mkfs.ext4 -L rootfs /dev/sde2
+mke2fs 1.44.6 (5-Mar-2019)
+/dev/sde2 contains a ext4 file system labelled 'rootfs'
+        created on Thu Aug  8 09:43:13 2019
+Proceed anyway? (y,N) y
+Creating filesystem with 245504 4k blocks and 61440 inodes
+Filesystem UUID: 7ce1bfb2-4862-4dea-82a3-e5b782115d7b
+Superblock backups stored on blocks: 
+        32768, 98304, 163840, 229376
+
+Allocating group tables: done
+Writing inode tables: done
+Creating journal (4096 blocks): done
+Writing superblocks and filesystem accounting information: done
+```
+_note: command line adds rootfs label_
+
+5. Mount file systems
+
+```
+$ sudo mount /dev/sde1 /mnt/rpi
+$ sudo mount /dev/sde2 /mnt/rootfs
+```
+
+## Copy Raspberry Pi firmwares
+
+1. Clone firmware repository
+
+```
+$ git clone git@github.com:raspberrypi/firmware.git
+```
+
+2. Copy files to fat32 partition
+
+```
+$ cd firmware
+$ cp -a boot/start* /mnt/rpi/.
+$ cp -a boot/fixup* /mnt/rpi/.
+$ cp -a boot/overlays/ /mnt/rpi/.
```

Comment:
`/mnt/rpi` is vFAT. So, `-a` should not be used. Also, sudo is necessary.
```
$ sudo cp boot/start* /mnt/rpi/.
$ sudo cp boot/fixup* /mnt/rpi/.
$ sudo cp -r boot/overlays /mnt/rpi/.
```

Reply:
Also bootcode.bin is forgotten.

Reply:
Right! I added bootcode.bin.